crates/bevy_ecs/src/traversal.rs

```diff
@@ -0,0 +1,33 @@
+//! A trait for components that let you traverse the ECS
+
+use crate::{
+    component::{Component, StorageType},
+    entity::Entity,
+};
+
+/// A component that holds a pointer to another entity.
+///
+/// The implementor is responsible for ensuring that `Traversal::next` cannot produce infinite loops.
+pub trait Traversal: Component {
+    /// Returns the next entity to visit.
+    fn next(&self) -> Option<Entity>;
```

Comment:
I experimented with several versions of this api and ultimately decided to implement the minimum necessary for the picking port.

Allowing multiple outputs, either with a `&mut Self` or a `&[Entity]` ended up significantly complicating observer dispatch because it requires a branching stack of dispatched events, where traversal could be halted at any point.

Which is to say, I'd really like to leave that sort of thing to follow-up.